File: java/src/test/java/org/libpetri/export/DotExporterTest.java
package org.libpetri.export;

import org.junit.jupiter.api.Test;
import org.libpetri.core.*;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.export.graph.RankDir;

import java.time.Duration;
import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class DotExporterTest {

    private static final Place<String> START = Place.of("Start", String.class);
    private static final Place<String> END = Place.of("End", String.class);

    @Test
    void producesValidDotForSimpleNet() {
        var t = Transition.builder("Process")
            .inputs(In.one(START))
            .outputs(Out.place(END))
            .build();
        var net = PetriNet.builder("SimpleNet").transition(t).build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("digraph SimpleNet {"));
        assertTrue(dot.contains("rankdir=TB;"));
        assertTrue(dot.contains("p_Start ["));
        assertTrue(dot.contains("p_End ["));
        assertTrue(dot.contains("t_Process ["));
        assertTrue(dot.contains("p_Start -> t_Process"));
        assertTrue(dot.contains("t_Process -> p_End"));
        assertTrue(dot.contains("}"));
    }

    @Test
    void containsCircleShapesForPlacesAndBoxForTransitions() {
        var t = Transition.builder("T")
            .inputs(In.one(START))
            .outputs(Out.place(END))
            .build();
        var net = PetriNet.builder("Test").transition(t).build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("shape=\"circle\""));
        assertTrue(dot.contains("shape=\"box\""));
    }

    @Test
    void handlesAllFiveArcTypes() {
        var input = Place.of("Input", String.class);
        var output = Place.of("Output", String.class);
        var blocker = Place.of("Blocker", String.class);
        var config = Place.of("Config", String.class);
        var cache = Place.of("Cache", String.class);

        var t = Transition.builder("Process")
            .inputs(In.one(input))
            .outputs(Out.place(output))
            .inhibitor(blocker)
            .read(config)
            .reset(cache)
            .build();
        var net = PetriNet.builder("AllArcs").transition(t).build();

        var dot = DotExporter.export(net);

        // Input arc
        assertTrue(dot.contains("p_Input -> t_Process"));
        // Output arc
        assertTrue(dot.contains("t_Process -> p_Output"));
        // Inhibitor arc (odot)
        assertTrue(dot.contains("p_Blocker -> t_Process"));
        assertTrue(dot.contains("arrowhead=\"odot\""));
        // Read arc (dashed)
        assertTrue(dot.contains("p_Config -> t_Process"));
        assertTrue(dot.contains("style=\"dashed\""));
        // Reset arc (bold)
        assertTrue(dot.contains("t_Process -> p_Cache"));
        assertTrue(dot.contains("style=\"bold\""));
    }

    @Test
    void handlesXorOutputsWithBranchLabels() {
        var success = Place.of("Success", String.class);
        var error = Place.of("Error", String.class);
        var t = Transition.builder("Process")
            .inputs(In.one(START))
            .outputs(Out.xor(success, error))
            .build();
        var net = PetriNet.builder("XorNet").transition(t).build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("label=\"Success\""));
        assertTrue(dot.contains("label=\"Error\""));
        // XOR transitions use standard box shape
        assertTrue(dot.contains("shape=\"box\""));
    }

    @Test
    void handlesWeightedInputArcs() {
        var t = Transition.builder("Batch")
            .inputs(In.exactly(3, START))
            .outputs(Out.place(END))
            .build();
        var net = PetriNet.builder("Test").transition(t).build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("\u00d73"));
    }

    @Test
    void includesTimingInTransitionLabels() {
        var t = Transition.builder("Process")
            .inputs(In.one(START))
            .outputs(Out.place(END))
            .timing(Timing.delayed(Duration.ofMillis(500)))
            .build();
        var net = PetriNet.builder("Test").transition(t).build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("[500"));
    }

    @Test
    void respectsConfigDirection() {
        var t = Transition.builder("T")
            .inputs(In.one(START))
            .outputs(Out.place(END))
            .build();
        var net = PetriNet.builder("Test").transition(t).build();

        var dot = DotExporter.export(net, ExportConfig.leftToRight());

        assertTrue(dot.contains("rankdir=LR;"));
    }

    @Test
    void rendersEnvironmentPlacesWithDashedStyle() {
        var envPlace = Place.of("Events", String.class);
        var t = Transition.builder("T")
            .inputs(In.one(envPlace))
            .outputs(Out.place(END))
            .build();
        var net = PetriNet.builder("Test").transition(t).build();

        var config = new ExportConfig(RankDir.TB, true, true, true, Set.of("Events"));
        var dot = DotExporter.export(net, config);

        assertTrue(dot.contains("style=\"filled,dashed\""));
        assertTrue(dot.contains("fillcolor=\"#f8d7da\""));
    }

    @Test
    void handlesMultiTransitionNet() {
        var pending = Place.of("Pending", String.class);
        var validated = Place.of("Validated", String.class);
        var processed = Place.of("Processed", String.class);

        var validate = Transition.builder("Validate")
            .inputs(In.one(pending))
            .outputs(Out.place(validated))
            .build();
        var process = Transition.builder("Process")
            .inputs(In.one(validated))
            .outputs(Out.place(processed))
            .build();

        var net = PetriNet.builder("Pipeline")
            .transitions(validate, process)
            .build();

        var dot = DotExporter.export(net);

        assertTrue(dot.contains("p_Pending"));
        assertTrue(dot.contains("p_Validated"));
        assertTrue(dot.contains("p_Processed"));
        assertTrue(dot.contains("t_Validate"));
        assertTrue(dot.contains("t_Process"));
    }

    @Test
    void sanitizesNames() {
        assertEquals("hello_world", DotExporter.sanitize("hello_world"));
        assertEquals("my_place_name", DotExporter.sanitize("my-place.name"));
        assertEquals("Place_Name", DotExporter.sanitize("Place Name"));
    }
}

File: typescript/tests/export/dot-exporter.test.ts
import { describe, it, expect } from 'vitest';
import { dotExport } from '../../src/export/dot-exporter.js';
import { PetriNet } from '../../src/core/petri-net.js';
import { Transition } from '../../src/core/transition.js';
import { place } from '../../src/core/place.js';
import { one, exactly } from '../../src/core/in.js';
import { outPlace, andPlaces, xorPlaces, timeout } from '../../src/core/out.js';
import { delayed } from '../../src/core/timing.js';

describe('dotExport', () => {
  it('produces valid DOT for a simple net', () => {
    const p1 = place('Start');
    const p2 = place('End');
    const t = Transition.builder('Process')
      .inputs(one(p1))
      .outputs(outPlace(p2))
      .build();
    const net = PetriNet.builder('SimpleNet').transition(t).build();

    const dot = dotExport(net);

    expect(dot).toContain('digraph SimpleNet {');
    expect(dot).toContain('rankdir=TB;');
    expect(dot).toContain('p_Start [');
    expect(dot).toContain('p_End [');
    expect(dot).toContain('t_Process [');
    expect(dot).toContain('p_Start -> t_Process');
    expect(dot).toContain('t_Process -> p_End');
    expect(dot).toContain('}');
  });

  it('contains circle shapes for places and box for transitions', () => {
    const p1 = place('In');
    const p2 = place('Out');
    const t = Transition.builder('T')
      .inputs(one(p1))
      .outputs(outPlace(p2))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net);

    // Places are circles with xlabel and fixedsize
    expect(dot).toMatch(/p_In \[.*shape="circle"/);
    expect(dot).toMatch(/p_In \[.*label=""/);
    expect(dot).toMatch(/p_In \[.*xlabel="In"/);
    expect(dot).toMatch(/p_In \[.*fixedsize="true"/);
    expect(dot).toMatch(/p_Out \[.*shape="doublecircle"/);
    // Transitions are boxes
    expect(dot).toMatch(/t_T \[.*shape="box"/);
  });

  it('handles a net with all 5 arc types', () => {
    const input = place('Input');
    const output = place('Output');
    const blocker = place('Blocker');
    const config = place('Config');
    const cache = place('Cache');

    const t = Transition.builder('Process')
      .inputs(one(input))
      .outputs(outPlace(output))
      .inhibitor(blocker)
      .read(config)
      .reset(cache)
      .build();
    const net = PetriNet.builder('AllArcs').transition(t).build();

    const dot = dotExport(net);

    // Input arc
    expect(dot).toContain('p_Input -> t_Process');
    // Output arc
    expect(dot).toContain('t_Process -> p_Output');
    // Inhibitor arc (odot arrowhead)
    expect(dot).toMatch(/p_Blocker -> t_Process.*arrowhead="odot"/);
    // Read arc (dashed)
    expect(dot).toMatch(/p_Config -> t_Process.*style="dashed"/);
    // Reset arc (bold)
    expect(dot).toMatch(/t_Process -> p_Cache.*style="bold"/);
  });

  it('handles XOR outputs with branch labels', () => {
    const start = place('Start');
    const success = place('Success');
    const error = place('Error');
    const t = Transition.builder('Process')
      .inputs(one(start))
      .outputs(xorPlaces(success, error))
      .build();
    const net = PetriNet.builder('XorNet').transition(t).build();

    const dot = dotExport(net);

    expect(dot).toContain('label="Success"');
    expect(dot).toContain('label="Error"');
    // XOR transitions use standard box shape
    expect(dot).toMatch(/t_Process \[.*shape="box"/);
  });

  it('handles weighted input arcs', () => {
    const start = place('Start');
    const end = place('End');
    const t = Transition.builder('Batch')
      .inputs(exactly(3, start))
      .outputs(outPlace(end))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net);

    expect(dot).toContain('\u00d73');
  });

  it('includes timing in transition labels', () => {
    const p1 = place('Start');
    const p2 = place('End');
    const t = Transition.builder('Process')
      .inputs(one(p1))
      .outputs(outPlace(p2))
      .timing(delayed(500))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net);

    expect(dot).toContain('[500');
  });

  it('handles timeout outputs', () => {
    const start = place('Start');
    const timeoutP = place('Timeout');
    const t = Transition.builder('Process')
      .inputs(one(start))
      .outputs(timeout(5000, outPlace(timeoutP)))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net);

    expect(dot).toContain('5000ms');
  });

  it('respects config direction', () => {
    const p1 = place('Start');
    const p2 = place('End');
    const t = Transition.builder('T')
      .inputs(one(p1))
      .outputs(outPlace(p2))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net, { direction: 'LR', showTypes: true, showIntervals: true, showPriority: true });

    expect(dot).toContain('rankdir=LR;');
  });

  it('renders environment places with dashed style', () => {
    const envPlace = place('Events');
    const output = place('Out');
    const t = Transition.builder('T')
      .inputs(one(envPlace))
      .outputs(outPlace(output))
      .build();
    const net = PetriNet.builder('Test').transition(t).build();

    const dot = dotExport(net, {
      direction: 'TB',
      showTypes: true,
      showIntervals: true,
      showPriority: true,
      environmentPlaces: new Set(['Events']),
    });

    expect(dot).toMatch(/p_Events \[.*style="filled,dashed"/);
    expect(dot).toMatch(/p_Events \[.*fillcolor="#f8d7da"/);
  });

  it('handles multi-transition net', () => {
    const pending = place('Pending');
    const validated = place('Validated');
    const processed = place('Processed');

    const validate = Transition.builder('Validate')
      .inputs(one(pending))
      .outputs(outPlace(validated))
      .build();
    const process = Transition.builder('Process')
      .inputs(one(validated))
      .outputs(outPlace(processed))
      .build();

    const net = PetriNet.builder('Pipeline')
      .transitions(validate, process)
      .build();

    const dot = dotExport(net);

    expect(dot).toContain('p_Pending');
    expect(dot).toContain('p_Validated');
    expect(dot).toContain('p_Processed');
    expect(dot).toContain('t_Validate');
    expect(dot).toContain('t_Process');
  });
});
